### ISSUE TITLE
weight in frozen .pb file from tensorflow seem are not converted to .mlmodel

### ISSUE THREAD

**Issue by ChiFang** [tf-coreml/tf-coreml#95]
my environment:
-------------------------
OS: Ubuntu 16
Python: 2.7.12
Package:
coremltools (0.7)
numpy (1.13.3)
protobuf (3.5.0.post1)
six (1.10.0)
tensorflow-gpu (1.2.0)
tfcoreml (0.1.0)
-------------------------
here is part of my net structure:
```
def My_CustomLayer(net, train=True, layer_id = ''):
    batch, rows, cols, channels = [i.value for i in net.get_shape()]
    var_shape = [channels]
    mu, sigma_sq = tf.nn.moments(net, [1,2], keep_dims=True)
    shift = tf.Variable(tf.zeros(var_shape), name = 'norm_shift_' + layer_id)
    scale = tf.Variable(tf.ones(var_shape), name = 'norm_scale_' + layer_id)
    epsilon = 1e-3
    normalized = (net-mu)* tf.rsqrt(sigma_sq + epsilon)
    return scale * normalized + shift
```

Shift and scale are trainable variable in tensorflow and .pb work perfectly when I predict.
The converting from .pb to .mlmodel do not show any error.

But when I use .mlmodel to predict, it will return totally wrong result.

once I remove shift and scale it will work fine when I use .mlmodel.
(but I need shift and scale to improve my net accuracy and training speed.....)

It seem the weight  of shift and scale are not converted to .mlmodel.

any suggestion?  Thanks~~~

**Comment by aseemw** [comment]
Hi @ChiFang, This could be a bug in the converter. We will take a look. Thank you for the code snippet. 

Meanwhile, if you want to continue your own investigation you can use the coremltools [visualization](https://github.com/apple/coremltools/blob/2bcaa8ef72e4f7b922e739207547d784bb2e6622/coremltools/models/utils.py#L781) tool to analyze the .mlmodel.

**Comment by ChiFang** [comment]
Hi @aseemw  I trace the log info during converting, I found following msg:

> Automatic shape interpretation succeeded for input blob img_placeholder:0
> 5/481: Converting op name: norm_shift_norm_9/read ( type:  Identity )
> 10/481: Converting op name: conv9/read ( type:  Identity )
> 13/481: Converting op name: norm_shift_norm_8/read ( type:  Identity )
> 19/481: Converting op name: conv8/read ( type:  Identity )
> 22/481: Converting op name: norm_shift_norm_7/read ( type:  Identity )
> 28/481: Converting op name: conv7/read ( type:  Identity )
> 31/481: Converting op name: norm_shift_norm_6/read ( type:  Identity )
> 36/481: Converting op name: conv6/read ( type:  Identity )
> 39/481: Converting op name: norm_shift_5_3/read ( type:  Identity )
> 45/481: Converting op name: conv5_3/read ( type:  Identity )
> 48/481: Converting op name: norm_shift_5_2/read ( type:  Identity )
> 54/481: Converting op name: conv5_2/read ( type:  Identity )
> 57/481: Converting op name: norm_shift_5_1/read ( type:  Identity )
> 63/481: Converting op name: conv5_1/read ( type:  Identity )
> 66/481: Converting op name: norm_shift_5_0/read ( type:  Identity )
> 71/481: Converting op name: conv5_0/read ( type:  Identity )
> 74/481: Converting op name: norm_shift_4_3/read ( type:  Identity )
> 80/481: Converting op name: conv4_3/read ( type:  Identity )
> 83/481: Converting op name: norm_shift_4_2/read ( type:  Identity )
> 89/481: Converting op name: conv4_2/read ( type:  Identity )
> 92/481: Converting op name: norm_shift_4_1/read ( type:  Identity )
> 98/481: Converting op name: conv4_1/read ( type:  Identity )
> 101/481: Converting op name: norm_shift_4_0/read ( type:  Identity )
> 106/481: Converting op name: conv4_0/read ( type:  Identity )
> 109/481: Converting op name: norm_shift_3_3/read ( type:  Identity )
> 115/481: Converting op name: conv3_3/read ( type:  Identity )
> 118/481: Converting op name: norm_shift_3_2/read ( type:  Identity )
> 124/481: Converting op name: conv3_2/read ( type:  Identity )
> 127/481: Converting op name: norm_shift_3_1/read ( type:  Identity )
> 133/481: Converting op name: conv3_1/read ( type:  Identity )
> 136/481: Converting op name: norm_shift_3_0/read ( type:  Identity )
> 141/481: Converting op name: conv3_0/read ( type:  Identity )
> 144/481: Converting op name: norm_shift_norm_2/read ( type:  Identity )
> 149/481: Converting op name: conv2/read ( type:  Identity )
> 152/481: Converting op name: norm_shift_norm_1/read ( type:  Identity )
> 157/481: Converting op name: conv1/read ( type:  Identity )
> 160/481: Converting op name: norm_shift_norm_0/read ( type:  Identity )
> 165/481: Converting op name: conv0/read ( type:  Identity )
> 166/481: Converting op name: img_placeholder ( type:  Placeholder )
> Skipping name of placeholder

is that mean "norm_shift_norm" and "conv" are by pass (or skip) by the converter?

**Comment by slin07** [comment]
Hi, @ChiFang, I made a small script that tries to reproduce your problem, but it seems to be working properly. See below:
```
import numpy as np
import tensorflow as tf
from tensorflow.python.tools.freeze_graph import freeze_graph

def My_CustomLayer(net, train=True, layer_id = ''):
    batch, rows, cols, channels = [i.value for i in net.get_shape()]
    var_shape = [channels]
    mu, sigma_sq = tf.nn.moments(net, [1,2], keep_dims=True)
    shift = tf.Variable(tf.constant(np.random.rand(*var_shape).astype('float32')), name = 'norm_shift_' + layer_id)
    scale = tf.Variable(tf.constant(np.random.rand(*var_shape).astype('float32')), name = 'norm_scale_' + layer_id)
    epsilon = 1e-3
    normalized = (net-mu)* tf.rsqrt(sigma_sq + epsilon)
    return scale * normalized + shift

img_input = tf.placeholder(tf.float32, [None, 10, 10, 3], name='input')
net = My_CustomLayer(img_input, True, '1')

init = tf.global_variables_initializer()
saver = tf.train.Saver()

model_dir = './'
graph_def_file = './tf_graph.pb'
checkpoint_file = './tf_model.ckpt'
frozen_model_file = './tf_frozen.pb'
mlmodel_path = './model.mlmodel'

in_array = np.random.rand(1,10,10,3)

with tf.Session() as sess:
    sess.run(init)
    # initialize
    sess.run(tf.global_variables_initializer())
    # prepare the tensorflow inputs
    feed_dict = {img_input : in_array}
    result = sess.run(net, feed_dict=feed_dict)
    # save graph definition somewhere
    tf.train.write_graph(sess.graph, model_dir, graph_def_file)
    # save the weights
    saver.save(sess, checkpoint_file)
    
output_node_names = net.op.name

freeze_graph(input_graph=graph_def_file,
             input_saver="",
             input_binary=False,
             input_checkpoint=checkpoint_file,
             output_node_names=output_node_names,
             restore_op_name="save/restore_all",
             filename_tensor_name="save/Const:0",
             output_graph=frozen_model_file,
             clear_devices=True,
             initializer_nodes="")
             
output_tensor_names = net.op.name + ':0'

import tfcoreml
input_name_shape_dict = {img_input.name: [1,10,10,3]}
mlmodel = tfcoreml.convert(tf_model_path=frozen_model_file,
                        mlmodel_path=mlmodel_path,
                        output_feature_names=[output_tensor_names],
                        input_name_shape_dict=input_name_shape_dict)

coreml_in_array = in_array.squeeze().transpose((2,0,1))
coreml_input = {'input__0' : coreml_in_array}
coreml_pred = mlmodel.predict(coreml_input)

tf_res_transposed = result.squeeze().transpose((2,0,1)).flatten()
cm_res = coreml_pred['add_1__0'].flatten()

err = tf_res_transposed - cm_res
print('Max error = ', np.max(np.abs(err)))
```
The output of this script is:
```
('Max error = ', 1.33514404296875e-05)
```

**Comment by ChiFang** [comment]
thanks @slin07.  I follow your code to verify my network with python on Mac and it work fine. The root cause is that IOS engineer use wrong input format on iphone....

@aseemw thanks for the great project. I am so surprised that tfcoreml support tensorflow more than tensorflow lite....
